File: src/main/kotlin/me/pcasaes/hexoids/infrastructure/health/CapacityHealthCheck.java
package me.pcasaes.hexoids.infrastructure.health;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import me.pcasaes.hexoids.infrastructure.disruptor.QueueMetric;
import org.eclipse.microprofile.health.HealthCheck;
import org.eclipse.microprofile.health.HealthCheckResponse;
import org.eclipse.microprofile.health.Liveness;

import java.util.List;

@Liveness
@ApplicationScoped
public class CapacityHealthCheck implements HealthCheck {

    private final List<QueueMetric> queueMetricList;

    @Inject
    public CapacityHealthCheck(List<QueueMetric> queueMetricList) {
        this.queueMetricList = queueMetricList;
    }

    @Override
    public HealthCheckResponse call() {
        return queueMetricList
                .stream()
                .sorted(QueueMetric::compare)
                .filter(QueueMetric::isOverCapacity)
                .map(m -> "Service is over capacity. Queue '" + m.getName() + "' has load factory " + m.getLoadFactor())
                .map(HealthCheckResponse::down)
                .findFirst()
                .orElse(HealthCheckResponse.up("Plenty of capacity"));
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/infrastructure/health/CapacityHealthCheck.kt
package me.pcasaes.hexoids.infrastructure.health

import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import me.pcasaes.hexoids.infrastructure.disruptor.QueueMetric
import org.eclipse.microprofile.health.HealthCheck
import org.eclipse.microprofile.health.HealthCheckResponse
import org.eclipse.microprofile.health.Liveness

@Liveness
@ApplicationScoped
class CapacityHealthCheck @Inject constructor(private val queueMetricList: List<QueueMetric>) : HealthCheck {
    override fun call(): HealthCheckResponse {
        return queueMetricList
            .stream()
            .sorted { a, b -> QueueMetric.Companion.compare(a, b) }
            .filter { obj -> obj.isOverCapacity() }
            .map { m -> "Service is over capacity. Queue '${m.getName()}' has load factory ${m.getLoadFactor()}" }
            .map { name -> HealthCheckResponse.down(name) }
            .findFirst()
            .orElse(HealthCheckResponse.up("Plenty of capacity"))
    }
}
